### PR TITLE
HDDS-7884. ofs: Recursive delete of path of symlink with empty source fails

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

For link bucket as orphan, fix the failure of volume not found or bucket not found.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7884

## How was this patch tested?

Integration test added for the orphan link bucket deletion

### CODE REVIEW

**Comment by neils-dev** [comment]
Thanks @sumitagrawl for this PR and for your comments on the jira.  Pls note, looks like an integration test is failing when testing the sym links in CI workflow:
`TestOzoneFileSystemWithLinks.testLoopInLinkBuckets:228 Should throw Exception due to loop in Link Buckets`

**Comment by kerneltime** [comment]
@SaketaChalamchala @tanvipenumudy can you please take a look?

**Comment by neils-dev** [comment]
Hi @sumitagrawl , I saw a problem with the symlink delete problem with orphaned link while reviewing this PR.  It occurs when the src volume/bucket for the link is deleted.  When the src for the link is deleted it causes an orphaned bucket link.  This in turn currently causes an error when the target link volume/bucket is recursively listed.  This should be handled.  It can be reproduced by,

_create src and target link._
`ozone sh bucket link /vol1/abc /s3v/abc`
_delete src volume/bucket_
`ozone fs -rm -R -skipTrash ofs://om/vol1`

_list recursively the target link:_
`ozone fs -ls -R ofs://om/s3v`
**drwxrwxrwx   - hadoop hadoop          0 2023-03-20 19:20 ofs://localhost/s3v/abc
ls: Volume not found when getting bucket info**

**Comment by sumitagrawl** [comment]
> Hi @sumitagrawl , I saw a problem with the symlink delete problem with orphaned link while reviewing this PR. It occurs when the src volume/bucket for the link is deleted. When the src for the link is deleted it causes an orphaned bucket link. This in turn currently causes an error when the target link volume/bucket is recursively listed. This should be handled. It can be reproduced by,
> 
> _create src and target link._ `ozone sh bucket link /vol1/abc /s3v/abc` _delete src volume/bucket_ `ozone fs -rm -R -skipTrash ofs://om/vol1`
> 
> _list recursively the target link:_ `ozone fs -ls -R ofs://om/s3v` **drwxrwxrwx - hadoop hadoop 0 2023-03-20 19:20 ofs://localhost/s3v/abc ls: Volume not found when getting bucket info**

@neils-dev Above case is handled and PR is updated

**Comment by smengcl** [comment]
@neils-dev Looks like [the test case failure](https://github.com/apache/ozone/pull/4405#issuecomment-1471145558) is also related to #4246 ?

**Comment by neils-dev** [comment]
> @neils-dev Looks like https://github.com/apache/ozone/pull/4405#issuecomment-1471145558 is also related to https://github.com/apache/ozone/pull/4246 ?

Thanks @smengcl .  Actually, the failure is something that should be handled here.  The other PR we have open, [#4246](https://github.com/apache/ozone/pull/4246) is for recursive deletes with symlinks where the src bucket _is_ intact, however in this case the src bucket is _removed_ leaving an orphan symlink.  Failures occurring with operations on this orphan symlink are handled by Sumit's PR.

**Review by sadanand48 — COMMENTED** [review]
Thanks @sumitagrawl for working on this. Overall change looks good. left few minor comments.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystemWithLinks.java [hunk 1]
```diff
@@ -225,11 +225,9 @@ public void testLoopInLinkBuckets() throws Exception {
 
     try {
       FileSystem.get(conf);
-      Assert.fail("Should throw Exception due to loop in Link Buckets");
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystemWithLinks.java:228]: @sadanand48 Thanks for review, as part of this PR, delete and ls is fixed. rename is not allowed for bucket. other file/directory operation will fail as source bucket is not present, which is expected.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java [hunk 2]
```diff
@@ -69,9 +71,21 @@ public static BucketLayout resolveLinkBucketLayout(OzoneBucket bucket,
...
+      try {
+        sourceBucket =
+            objectStore.getVolume(bucket.getSourceVolume())
+                .getBucket(bucket.getSourceBucket());
+      } catch (OMException ex) {
+        if (ex.getResult().equals(VOLUME_NOT_FOUND)
+            || ex.getResult().equals(BUCKET_NOT_FOUND)) {
+          // for orphan link bucket, return layout as link bucket
+          bucket.setSourcePathExist(false);
+          return bucket.getBucketLayout();
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:87]: @sadanand48 Updated...

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java [hunk 3]
```diff
@@ -888,8 +889,14 @@ public List<FileStatusAdapter> listStatus(String pathStr, boolean recursive,
...
     try {
       OzoneBucket bucket = getBucket(ofsPath, false);
-      List<OzoneFileStatus> statuses = bucket
-          .listStatus(keyName, recursive, startKey, numEntries);
+      List<OzoneFileStatus> statuses;
+      if (bucket.isSourcePathExist()) {
+        statuses = bucket
+            .listStatus(keyName, recursive, startKey, numEntries);
+      } else {
+        statuses = Collections.emptyList();
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:899]: @sadanand48 Updated...

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java [hunk 4]
```diff
@@ -720,6 +670,99 @@ public boolean delete(Path f, boolean recursive) throws IOException {
...
+  }
+
+  private boolean isLinkBucket(Path f, OFSPath ofsPath) {
+    try {
+      OzoneBucket bucket = adapterImpl.getBucket(ofsPath, false);
+      if (bucket.isLink()) {
+        return true;
+      }
+    } catch (Exception ex) {
+      return false;
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java:711]: @sadanand48 Updated...

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java [hunk 5]
```diff
@@ -720,6 +670,99 @@ public boolean delete(Path f, boolean recursive) throws IOException {
     return result;
   }
 
+  private boolean deleteBucket(Path f, boolean recursive, OFSPath ofsPath)
+      throws IOException {
+    // check status of normal bucket
+    try {
+      getFileStatus(f);
+    } catch (FileNotFoundException ex) {
+      // remove orpahan link bucket directly
```
[the remark this answers is not in the document]
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java]: Updated...

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java [hunk 6]
```diff
@@ -657,8 +657,9 @@ private FileStatusAdapter getFileStatusForKeyOrSnapshot(
       OFSPath ofsPath, URI uri, Path qualifiedPath, String userName)
       throws IOException {
     String key = ofsPath.getKeyName();
+    OzoneBucket bucket = null;
```
Comment (neils-dev) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:660]: Nit: unnecessary change.  Variable `bucket` only used in scope of try block and can remain declared there.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java [hunk 7]
```diff
@@ -720,6 +670,101 @@ public boolean delete(Path f, boolean recursive) throws IOException {
...
   }
 
+  private boolean deleteBucket(Path f, boolean recursive, OFSPath ofsPath)
+      throws IOException {
+    // check status of normal bucket
+    try {
+      getFileStatus(f);
+    } catch (FileNotFoundException ex) {
+      // remove orphan link bucket directly
+      if (isLinkBucket(f, ofsPath)) {
```
Comment (neils-dev) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java:680]: Within this catch block we can just check `if isLinkBucket == false` and return `LOG.warn(Path does not exist)` with `false`.  Outside the catch block the bucket is checked if link, if true, then the orphan bucket is removed from volume and returned.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java [hunk 8]
```diff
@@ -647,6 +647,68 @@ public void testDeleteToTrashOrSkipTrash() throws Exception {
     }
   }
 
+  @Test
```
Comment (neils-dev) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:650]: Can this test be placed in the integration test suite, `TestRootedOzoneFileSystem`?  Grouped with symlink bucket tests as,
https://github.com/apache/ozone/blob/0c84ff1cdb5208e4932bcc6195b26fcaaeece22b/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java#L1520

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java [hunk 9]
```diff
@@ -69,9 +71,24 @@ public static BucketLayout resolveLinkBucketLayout(OzoneBucket bucket,
...
+                .getBucket(bucket.getSourceBucket());
+      } catch (OMException ex) {
+        if (ex.getResult().equals(VOLUME_NOT_FOUND)
+            || ex.getResult().equals(BUCKET_NOT_FOUND)) {
+          // for orphan link bucket, return layout as link bucket
+          bucket.setSourcePathExist(false);
+          LOG.error("Source Bucket is not found, its orphan bucket and " +
+              "used link bucket {} layout {}", bucket.getName(),
+              bucket.getBucketLayout());
+          return bucket.getBucketLayout();
```
Comment (neils-dev) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:87]: What is the layout of the linked bucket?  Is it the layout of the original source bucket or is it defaulted to some value?

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java (hunk 7 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java:680]: @neils-dev 
isLinkBucket performs remote call to OM. So checking first as false in exception, and again checking outside for link bucket bucket will have 2 remote call. So I think this is not required to be refactored for performance.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java (hunk 8 above)
Comment (sumitagrawl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:650]: @neils-dev Cluster initialisation needs to be done with specific OM ID, and need some shell execute method to added again TestRootedOzoneFileSystem, and since this class is not doing shell related testing, so added in TestOzoneShellHA,
I think this is not required as way of testing the link bucket is different in both case.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java (hunk 9 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:87]: Its configured OM default layout "ozone.default.bucket.layout"

**Comment by neils-dev** [comment]
Thanks @sumitagrawl for reaching out.  Let me take a look and also let me coordinate with @smengcl having this update and [PR #4246](https://github.com/apache/ozone/pull/4246) added as they will overlap a bit and conflict when we merge.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java (hunk 7 above)
Comment (neils-dev) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java:680]: Thanks for pointing out the remote calls.  Makes sense.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java (hunk 9 above)
Comment (neils-dev) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:87]: Thanks.  I ran it the `OzoneClientUtils.resolveLinkBucketLayout` with default, and user specified layouts on bucket create.  No problems - the linked bucket object layout is set to the source bucket type.

**Comment by sumitagrawl** [comment]
> Thanks @sumitagrawl for reaching out. Let me take a look and also let me coordinate with @smengcl having this update and [PR #4246](https://github.com/apache/ozone/pull/4246) added as they will overlap a bit and conflict when we merge.

@neils-dev @smengcl Please check if the PR can be merged
